### PR TITLE
Faster RCNN script and .keras file

### PR DESCRIPTION
**Pull request by DataWithDino**
Bitte Reviewen <3

### CODE REVIEW

**Review by FredericKayser — CHANGES_REQUESTED** [review]
Could you remove the .keras file? 
Rest looks fine for a MVP, refactoring and compromising of functions relevant for both models can be done at a later point

**Review by FredericKayser — APPROVED** [review]

File: src/models/faster_rcnn_model.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        print(f"Test Loss: {test_loss}")
+
+    def _save_model(self):
+        """
+        Save the trained model to disk.
+        """
+        self.model.save('faster_rcnn_model.keras')
+
+
+if __name__ == '__main__':
```
Comment (simgae) [inline src/models/faster_rcnn_model.py:123]: Kannst du die main Methode mit der anderen main Methode in `__init__.py` mergen. Dann haben wir nur eine zentrale main Methode. Gerne kannst du auch das Training des YOLO Models dort aus kommentieren.

File: src/models/faster_rcnn_model.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+        self.model = Model(inputs=base_model.input, outputs=[rpn_cls, rpn_bbox])
+        # We have 2 outputs => supply 2 losses in the same order
+        self.model.compile(
+            optimizer='adam',
+            loss=['binary_crossentropy', 'mean_squared_error']
+        )
+
+    def train_model(self, epochs=2):
+        """
+        Train the model for a given number of epochs.
```
Comment (simgae) [inline src/models/faster_rcnn_model.py:95]: Kannst du den Parameter noch als doc-string hinzufügen?

File: src/models/faster_rcnn_model.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+            )
+
+
+class FasterRCNNModel:
+    """
+    Minimal demonstration of a Faster R-CNN–style model with two heads:
+      - A classification (rpn_cls) output
+      - A bounding box regression (rpn_reg) output
+
+    We supply matching label tuples (class_label, bbox_label) to avoid structure mismatch.
```
Comment (simgae) [inline src/models/faster_rcnn_model.py:54]: Evtl. hier noch die public Attribute der Klasse auflisten - dann sind wir einheitlicher.

File: src/models/faster_rcnn_model.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+from tensorflow.keras.callbacks import TensorBoard
+
+
+def preprocess_dataset(example):
+    """
+    Preprocess 'example' from the KITTI dataset and return (image, (class_label, bbox_label)).
+
+    For demonstration:
+    - We create a dummy 'class_label' (binary: 0) and 'bbox_label' (4 zeros).
+    - Real code should extract actual labels from 'example'.
```
Comment (simgae) [inline src/models/faster_rcnn_model.py:16]: Auch hier den Parameter in den doc-string aufnehmen.

**Review by simgae — COMMENTED** [review]
Alles nur Kleinigkeiten. Nach den Änderungen lgtm

File: src/models/faster_rcnn_model.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        base_model.trainable = False
+
+        # 4) Add flatten and two outputs:
+        x = base_model.output
+        x = Flatten()(x)
+
+        # Output 1: classification (binary, i.e. 1 neuron with sigmoid)
+        rpn_cls = Dense(1, activation='sigmoid', name='rpn_cls')(x)
+        # Output 2: bounding box coords (4 floats)
+        rpn_bbox = Dense(4, activation='linear', name='rpn_reg')(x)
```
Comment (FredericKayser) [inline src/models/faster_rcnn_model.py:83]: In the other model, we estimate 10 bounding boxes (for multi object detection) - can we do that as well here?

File: src/models/faster_rcnn_model.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+            input_shape=(128, 128, 3)
+        )
+        base_model.trainable = False
+
+        # 4) Add flatten and two outputs:
+        x = base_model.output
+        x = Flatten()(x)
+
+        # Output 1: classification (binary, i.e. 1 neuron with sigmoid)
+        rpn_cls = Dense(1, activation='sigmoid', name='rpn_cls')(x)
```
Comment (FredericKayser) [inline src/models/faster_rcnn_model.py:81]: I'm not sure if this approach will work as the sigmoid function is between 0 and 1. As far as I know you often use x neurons for x different objects and then use the softmax activation function to calculate the probability of each classification

Extension to 10 objects needs to be implemented as well

**Review by FredericKayser — CHANGES_REQUESTED** [review]

File: src/models/faster_rcnn_model.py [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+            include_top=False,
+            input_shape=(128, 128, 3)
+        )
+        base_model.trainable = False
+
+        # Add flatten and two output heads
+        x = base_model.output
+        x = Flatten()(x)
+
+        # Output 1: classification head for 10 objects using softmax activation.
```
Comment (FredericKayser) [inline src/models/faster_rcnn_model.py:101]: I think we need amount of objects * different classifications (so 10 * 5), so 50 different outputs for the classification

**Comment by simgae** [comment]
@DataWithDino hattest du einmal ausprobiert, was passiert wenn du ein Bild auswertest mit dem Model? Wie sind die Ergebnisse?

**Comment by simgae** [comment]
Due to performance issues this model architecture will not be implemented.
